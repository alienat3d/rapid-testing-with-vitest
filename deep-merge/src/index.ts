// 3.4 В самой же функции "deepMerge" мы используем для слияния двух объектов встроенный в JS метод объектов "assign".
// Go to [deep-merge\test\merge.test.ts]
// 3.7 Итак, добавим условие, чтобы проверить, является ли аргумент "a" массивом. И если является, то с помощью spread-оператора мы развернём содержимое первого и второго аргумента-массива в итоговый массив.
// 4.1 Теперь мы встречаем ошибку, это из-за того, что метод "assign" способен на поверхностное слияние, т.е. на верхнем уровне вложенности. Но мы можем сделать, что называется "shadow clone" объекта "a", а зачем пройтись по всем ключам объекта "b", используя цикл "for of". Чтобы получить предыдущее поведение мы могли бы напрямую присвоить значения объекта "b" итоговому "merged" объекту.
// Go to [deep-merge\test\deepMerge.test.ts]
// 4.4 Улучшим нашу реализацию, чтобы пройти также тест на глубокое слияние свойств объекта. Мы используем рекурсию здесь для вложенных в объекты свойств.
// 4.5 Отлично, тесты снова пройдены. Теперь у нас есть готовая функция для глубокого слияния двух объектов или массивов.
// Go to [deep-merge\test\deepMerge.test.ts]
// 5.4 Добавим ещё одну проверку для корректного вывода ошибки с помощью конструкта "throw new Error", в случае, если была зафиксирована попытка слияния разных типов данных.
export function deepMerge(a, b) {
  if (Array.isArray(a) && Array.isArray(b)) {
    return [...a, ...b];
  }
  if (Array.isArray(a) || Array.isArray(b) || typeof a !== typeof b) {
    throw new Error('Can not merge two different types.');
  }
  // return Object.assign(a, b);
  const merged = { ...a };

  for (const key of Object.keys(b)) {
    if (typeof a[key] === 'object' || Array.isArray(a[key])) {
      merged[key] = deepMerge(a[key], b[key]);
    } else merged[key] = b[key];
  }

  return merged;
}
