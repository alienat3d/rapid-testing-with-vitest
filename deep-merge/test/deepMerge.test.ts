import { expect, test } from 'vitest';
import { deepMerge } from '../src';

test('deep merge with props overlaps', () => {
  const merged = deepMerge(
    {
      name: 'Al',
      accounts: {
        github: 'alienat3d',
      },
    },
    {
      accounts: {
        twitter: 'alienat3d',
      },
    }
  );

  expect(merged).toEqual({
    name: 'Al',
    accounts: {
      github: 'alienat3d',
      twitter: 'alienat3d',
    },
  });
});

test('throws errors on merging two different types', () => {
  expect(() =>
    deepMerge(['foo', 'bar'], { foo: 'bar' }).toThrowError(
      'Error: Can not merge two different types.'
    )
  );
});

// * 7.0 На прошлом уроке мы научились работать со снэпшотами, с помощью которых удобно следить, что после внесения изменений в функцию результат её работы останется неизменным. Но порой мы хотим, чтобы этот результат после изменения функции изменился. Или мы можем изменить вводные данные для охвата разных сценариев и тогда нам нужно будет обновить также снэпшот соответственно. Рассмотрим два способа сделать это.
// 7.1 Добавим в первый объект свойство "languages", а во второй "typescript". Теперь результат функции отличается от того, который ожидался вначале. В консоли, во время работы теста, можно увидеть тот результат, на который нам нужно обновить
// 7.2 Первый способ обновить снэпшот это горячие клавиши. И сам Vitest подсказывает нам в консоли, что по нажатию на "u" можно обновить снэпшот. Но Vitest сделает это обновление единожды, а если мы изменим тест снова, то он снова провалится.
// 7.3 Другая опция обновить снэпшот это использовать CLI флаг. Мы можем написать в консоли "npx vitest -u" это запустит автоматическое обновление снэпшота, как только Vitest заметит несоответствие. Это полезно, если мы хотим переходить к следующему тесту быстро и получать быстрый ответ. Но, перед тем как использовать эту функцию, рекомендуется сохранить файл снэпшота куда-нибудь в другое место или сохранить состояние в Git.
test('deep merge with props overlaps', () => {
  const merged = deepMerge(
    {
      name: 'Al',
      accounts: {
        github: 'alienat3d',
      },
      languages: ['javascript'],
    },
    {
      accounts: {
        twitter: 'alienat3d',
      },
      languages: ['typescript', 'vue'],
    }
  );

  expect(merged).toMatchSnapshot();
});
