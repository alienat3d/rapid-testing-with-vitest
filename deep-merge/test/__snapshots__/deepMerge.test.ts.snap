// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`deep merge with props overlaps 1`] = `
{
  "accounts": {
    "github": "alienat3d",
    "twitter": "alienat3d",
  },
  "languages": [
    "javascript",
    "typescript",
    "vue",
  ],
  "name": "Al",
}
`;
